qiskit_experiments/analysis/curve_fitting.py

```diff
@@ -0,0 +1,291 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""
+Curve fitting functions for experiment analysis
+"""
+# pylint: disable = invalid-name
+
+from typing import List, Dict, Tuple, Callable, Optional, Union
+
+import numpy as np
+import scipy.optimize as opt
+from qiskit.exceptions import QiskitError
+from qiskit_experiments.base_analysis import AnalysisResult
+from qiskit_experiments.analysis.data_processing import filter_data
+
+
+def curve_fit(
+    func: Callable,
+    xdata: np.ndarray,
+    ydata: np.ndarray,
+    p0: Union[Dict[str, float], np.ndarray],
+    sigma: Optional[np.ndarray] = None,
+    bounds: Optional[Union[Dict[str, Tuple[float, float]], Tuple[np.ndarray, np.ndarray]]] = None,
+    **kwargs,
+) -> AnalysisResult:
+    r"""Perform a non-linear least squares to fit
+
+    This solves the optimization problem
+
+    .. math::
+        \Theta_{\mbox{opt}} = \arg\min_\Theta \sum_i
+            \sigma_i^{-2} (f(x_i, \Theta) -  y_i)^2
+
+    using ``scipy.optimize.curve_fit``.
+
+    Args:
+        func: a fit function `f(x, *params)`.
+        xdata: a 1D float array of x-data.
+        ydata: a 1D float array of y-data.
+        p0: initial guess for optimization parameters.
+        sigma: Optional, a 1D array of standard deviations in ydata.
+        bounds: Optional, lower and upper bounds for optimization
+                parameters.
+        kwargs: additional kwargs for scipy.optimize.curve_fit.
+
+    Returns:
+        result containing `popt` the optimal fit parameters,
+        `popt_err` the standard error estimates popt,
+        `pcov` the covariance matrix for the fit,
+        `chisq` the chi-squared parameter of fit,
+        `xrange` the range of xdata values used for fit.
+
+    Raises:
+        QiskitError: if the number of y-values in the dataset is not
+                     greater than the number of parameters in the
+                     fit function.
+    """
+    # Format p0 parameters if specified as dictionary
+    if isinstance(p0, dict):
+        param_keys = list(p0.keys())
```

Comment:
I think this will make the ordering of p0 as a vector unstable, which will probably mess someone up when they assume the covariance matrix is always ordered the same.  Maybe sort alphabetically by key or something?

Reply:
Currently the popt, popt_keys, popt_err, pcov ordering will be the ordering of `dict.keys`. The `popt_keys` is intended to remove any ambiguity for which params these are. I don't think it makes that much more sense to order alphabetically instead of the order the dict was originally constructed if p0 is a dict.